Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/BayesStretch.py

```diff
@@ -0,0 +1,332 @@
+                        Nspec=nsam,
+                        UnitX='',
+                        VerticalAxisUnit='MomentumTransfer',
+                        VerticalAxisValues=Qaxis)
+        unitx = mtd[fname + '_Sigma'].getAxis(0).setUnit("Label")
+        unitx.setLabel('sigma' , '')
+
+        CreateWorkspace(OutputWorkspace=fname + '_Beta',
+                        DataX=xBet,
+                        DataY=yBet,
+                        DataE=eBet,
+                        Nspec=nsam,
+                        UnitX='',
+                        VerticalAxisUnit='MomentumTransfer',
+                        VerticalAxisValues=Qaxis)
+        unitx = mtd[fname + '_Beta'].getAxis(0).setUnit("Label")
+        unitx.setLabel('beta' , '')
+
+        group = fname + '_Sigma,' + fname + '_Beta'
+        fit_ws = fname + '_Fit'
+        GroupWorkspaces(InputWorkspaces=group,
+                        OutputWorkspace=fit_ws)
+        contour_ws = fname + '_Contour'					
+        GroupWorkspaces(InputWorkspaces=groupZ,
+                        OutputWorkspace=contour_ws)
+
+        log_prog = Progress(self, start=0.8, end =1.0, nreports=8)
+        #Add some sample logs to the output workspaces
+        log_prog.report('Copying Logs to Fit workspace')
+        CopyLogs(InputWorkspace=self._sam_ws,
+                 OutputWorkspace=fit_ws)
+        log_prog.report('Adding Sample logs to Fit workspace')
+        QLAddSampleLogs(fit_ws, self._res_ws, prog, self._background, self._elastic, erange,
+                        (nbin, nrbin), self._resnormWS, self._wfile)
+        log_prog.report('Copying logs to Contour workspace')
+        CopyLogs(InputWorkspace=self._sam_ws,
+                 OutputWorkspace=contour_ws)
+        log_prog.report('Adding sample logs to Contour workspace')
+        QLAddSampleLogs(contour_ws, self._res_ws, prog, self._background, self._elastic, erange,
+                        (nbin, nrbin), self._resnormWS, self._wfile)
+        log_prog.report('Finialising log copying')
+
+        if self._save:
+            log_prog.report('Saving workspaces')
+            fit_path = os.path.join(workdir, fit_ws + '.nxs')
+            SaveNexusProcessed(InputWorkspace=fit_ws,
+                               Filename=fit_path)
+            logger.information('Output Fit file created : %s' % fit_path)
+            contour_path = os.path.join(workdir, contour_ws + '.nxs')                    # path name for nxs file
+            SaveNexusProcessed(InputWorkspace=contour_ws,
+                               Filename=contour_path)
+            logger.information('Output Contour file created : %s' % contour_path)
+            log_prog.report('Files Saved')
+
+        self.setProperty('OutputWorkspaceFit', fit_ws)
+        self.setProperty('OutputWorkspaceContour', contour_ws)
+        log_prog.report('Setting workspace properties')
+
+AlgorithmFactory.subscribe(BayesStretch)         # Register algorithm with Mantid
```

Comment:
@KarlPalmen I think you are right. I have refactored the PyExec to be shorter (and hopefully more readable). Let me know if you still think it's too long and I can give it another go.